src/System.Private.DataContractSerialization/src/System/Runtime/Serialization/AccessorBuilder.cs

```diff
@@ -0,0 +1,153 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Reflection;
+using System.Runtime.Serialization;
+using System.Text;
+using System.Threading.Tasks;
+
+namespace System.Runtime.Serialization
+{
+    internal static class FastInvokerBuilder
+    {
+        public delegate void Setter(ref object obj, object value);
+        public delegate object Getter(object obj);
+
+        private delegate void StructSetDelegate<T, T1>(ref T obj, T1 value);
+
+        private static MethodInfo s_buildGetAccessorInternal = typeof(FastInvokerBuilder).GetMethod(nameof(BuildGetAccessorInternal), BindingFlags.NonPublic | BindingFlags.Public | BindingFlags.Static);
+        private static MethodInfo s_buildSetAccessorInternal = typeof(FastInvokerBuilder).GetMethod(nameof(BuildSetAccessorInternal), BindingFlags.NonPublic | BindingFlags.Public | BindingFlags.Static);
+        private static MethodInfo s_make = typeof(FastInvokerBuilder).GetMethod(nameof(Make), BindingFlags.NonPublic | BindingFlags.Public | BindingFlags.Static);
+
+        public static Func<object> GetMakeNewInstanceFunc(Type type)
+        {
+            Func<object> make = (Func<object>)s_make.MakeGenericMethod(type).CreateDelegate(typeof(Func<object>));
+            return make;
+        }
+
+        public static Getter BuildGetAccessor(MemberInfo memberInfo)
+        {
+            if (memberInfo is PropertyInfo)
+            {
+                var propInfo = (PropertyInfo)memberInfo;
+                Func<PropertyInfo, Getter> buildGetAccessorGeneric = (Func<PropertyInfo, Getter>)s_buildGetAccessorInternal.MakeGenericMethod(propInfo.DeclaringType, propInfo.PropertyType).CreateDelegate(typeof(Func<PropertyInfo, Getter>));
+                Getter accessor = buildGetAccessorGeneric(propInfo);
+                return accessor;
+            }
+            else if (memberInfo is FieldInfo)
+            {
+                return (obj) =>
```

Comment:
I believe it might make a perf difference if we can avoid closures, like the one generated here for FieldInfo (because it needs to reference the memberInfo parameter).  If we changed the signature of Getter and Setter to take a MemberInfo parameter, I believe it would avoid the closure.  I *think* we could do this because DataMember.MemberInfo would be available to pass to the Getter or Setter.  @shmao -- could you take a look to see if it would be possible to pass memberInfo to these delegates?